File: mds.py
#!/usr/bin/env python3
"""
MDS API client.

Everything here was verified against production on 2026-08-07. The two
non-obvious things, both of which cost real time to find:

  1. `Authorization: <raw token>` — no "Bearer" prefix. Chrome's HAR export
     STRIPS Authorization headers, so this header is invisible in .har files
     and only appears in "Copy as cURL". Without it every endpoint returns an
     empty-body 401 from Spring Security, which looks like a bad token.

  2. /fis/auth/login returns HTTP 200 even on failure, with the error in
     `successStatus`. Data endpoints DO use a proper 401. So auth failure has
     to be detected two different ways depending on the endpoint.
"""
import json, os, subprocess, sys, time, urllib.parse
from datetime import datetime, timedelta

ROOT = os.path.dirname(os.path.abspath(__file__))
TOKEN_CACHE = os.path.join(ROOT, "data", ".token_cache.json")
TOKEN_TTL_S = 4 * 3600          # measured exactly 4.0h (iat -> exp)
REFRESH_MARGIN_S = 20 * 60      # refresh early so live requests never eat a 401


KEYS = ("MDS_BASE_URL", "MDS_USERNAME", "MDS_PASSWORD", "UI_PASSWORD",
        "APPS_SCRIPT_EXEC_URL", "APPS_SCRIPT_TOKEN", "METABASE_CARD_ID",
        "METABASE_DB_ID", "HISTORY_WINDOW_DAYS", "SWEEP_INTERVAL_MIN")


def load_env():
    """Config from a .env file when present, else from real environment
    variables. Container platforms (Fly, Render, Docker) inject secrets as env
    vars and have no .env, so requiring the file made the app undeployable
    anywhere but a VM. Real env vars win, so a platform secret can override a
    stale value baked into an image."""
    cfg = {}
    p = os.path.join(ROOT, ".env")
    if os.path.exists(p):
        cfg = dict(l.strip().split("=", 1) for l in open(p)
                   if "=" in l and not l.strip().startswith("#"))
    for k in KEYS:
        if os.environ.get(k):
            cfg[k] = os.environ[k]
    cfg.setdefault("MDS_BASE_URL", "https://fleet-green.moveinsync.com")
    if not cfg.get("MDS_USERNAME") or not cfg.get("MDS_PASSWORD"):
        sys.exit("MDS_USERNAME / MDS_PASSWORD not set (via .env or environment)")
    return cfg


class MDS:
    def __init__(self, env=None):
        self.env = env or load_env()
        self.base = self.env["MDS_BASE_URL"].rstrip("/")
        self._tok = None
        self._issued = 0
        self._load_cache()

    # ─────────────────────────── auth ───────────────────────────
    def _load_cache(self):
        try:
            c = json.load(open(TOKEN_CACHE))
            if time.time() - c["issued"] < TOKEN_TTL_S - REFRESH_MARGIN_S:
                self._tok, self._issued = c["tok"], c["issued"]
        except Exception:
            pass

    def _save_cache(self):
        os.makedirs(os.path.dirname(TOKEN_CACHE), exist_ok=True)
        json.dump({"tok": self._tok, "issued": self._issued}, open(TOKEN_CACHE, "w"))
        os.chmod(TOKEN_CACHE, 0o600)

    def login(self):
        body = {"username": self.env["MDS_USERNAME"], "password": self.env["MDS_PASSWORD"]}
        r = self._raw("POST", "/fis/auth/login", body, auth=False)
        # login signals failure in the BODY, not the status code
        if not r or not r.get("successStatus"):
            raise RuntimeError("MDS login failed: %s" % (r or {}).get("message"))
        d = r["data"]
        self._tok = {"Authorization": d["token"], "x-cds-token": d["token"],
                     "user_detail": d["userDetailToken"], "vendor_id": str(d["vendorId"])}
        self._issued = time.time()
        self._save_cache()
        return d

    def headers(self):
        if not self._tok or time.time() - self._issued > TOKEN_TTL_S - REFRESH_MARGIN_S:
            self.login()
        return dict(self._tok)

    # ─────────────────────────── transport ───────────────────────────
    def _raw(self, method, path, body=None, auth=True, timeout=90):
        cmd = ["curl", "-s", "-o", "/dev/stdout", "-w", "\n__HTTP__%{http_code}",
               "-X", method, "-H", "Accept: application/json, text/plain, */*",
               "-H", "Referer: %s/cds-green.html" % self.base]
        if auth:
            for k, v in self.headers().items():
                cmd += ["-H", "%s: %s" % (k, v)]
        if body is not None:
            cmd += ["-H", "Content-Type: application/json", "--data-raw", json.dumps(body)]
        cmd += ["--max-time", str(timeout), self.base + path]
        out = subprocess.run(cmd, capture_output=True, text=True).stdout
        payload, _, code = out.rpartition("__HTTP__")
        try:
            return json.loads(payload.strip() or "{}") if code.strip() != "401" else None
        except Exception:
            return None

    def call(self, method, path, body=None, _retried=False):
        """Data endpoints use a real 401 -> re-login once and retry."""
        r = self._raw(method, path, body)
        if r is None and not _retried:
            self.login()
            return self.call(method, path, body, _retried=True)
        return r

    # ─────────────────────────── endpoints ───────────────────────────
    def vendor_guids(self, buid):
        r = self.call("GET", "/fis/vendor/vendorowner/mapping/assigned")
        for m in r["data"]["vendorOwnerMappings"]:
            if m.get("buid") == buid:
                return [v["vendorGuid"] for v in m["assignedVendors"]]
        raise RuntimeError("no vendor mapping for %s" % buid)

    def shifts(self, buid, guids, day_ms):
        r = self.call("POST", "/fis/cds/shift/v3?nonShift=false&isPlannedTrip=false",
                      {"dates": [day_ms],
                       "buidsAndVendors": [{"businessUnit": buid, "vendorGuids": guids}]})
        out = []
        for d in (r or []):
            for b in d.get("businessUnitOfficeShifts", []):
                for o in b.get("officeShifts", []):
                    out.append(o)
        return out

    def trips(self, buid, guids, day_ms, offices=None):
        """Trip list. MDS caps this at 10 shifts per call, so batch.
        The 'shiftsByOfficeDirectionByBuid' shape is NOT a selectedShifts list —
        it's {buid: [{office:{name,id}, shiftsByDirection:{loginShifts,logoutShifts}}]}
        and `directions` must match whichever list is populated."""
        all_trips, seen = [], set()
        for off in self.shifts(buid, guids, day_ms):
            if offices and off["officeName"] not in offices:
                continue
            for direction, key in (("IN", "loginShifts"), ("OUT", "logoutShifts")):
                times = [s["shiftTime"] for s in (off.get(key) or [])]
                for i in range(0, len(times), 10):
                    chunk = times[i:i + 10]
                    body = {
                        "dateList": [day_ms, day_ms],
                        "businessUnitVendorsList": [{"businessUnit": buid, "vendorGuids": guids}],
                        "directions": [direction],
                        "shiftsByOfficeDirectionByBuid": {buid: [{
                            "office": {"name": off["officeName"], "id": off["officeId"]},
                            "shiftsByDirection": {
                                "loginShifts":  chunk if direction == "IN" else [],
                                "logoutShifts": [] if direction == "IN" else chunk}}]},
                        "selectedShifts": [{"buid": buid, "office": off["officeName"],
                                            "shift": t, "officeId": off["officeId"],
                                            "direction": direction} for t in chunk],
                        "plannedTrip": False}
                    r = self.call("POST",
                                  "/fis/cds/trip/filter?source=web&nonShift=false&cabUnassigned=false",
                                  body)
                    for t in ((r or {}).get("data") or []):
                        if t["tripId"] not in seen:
                            seen.add(t["tripId"]); all_trips.append(t)
        return all_trips

    def vehicles(self, buid, trip_guid):
        """Candidate pool. distance=true is REQUIRED: with distance=false,
        emptyLegInMetres is -1 for every cab."""
        g = urllib.parse.quote(trip_guid, safe="$-")
        r = self.call("GET", "/fis/cds/vendor/trips/%s/%s/vehicles?distance=true" % (buid, g))
        return (r or {}).get("data") or []

    @staticmethod
    def trip_guid(buid, trip_id):
        """Best-effort GUID construction, so a trip is reachable without trip/filter.
        Rule: strip dashes from the buid, chunk into 4s, join with dashes, then
        $ + zero-padded-10 trip id.
          ivycomptech-IVYHyd -> ivycomptechIVYHyd -> ivyc|ompt|echI|VYHy|d
          + 1135587 -> TRIPivyc-ompt-echI-VYHy-d$0001135587
        Verified for ivycomptech-IVYHyd only. Prefer the tripGuid returned by
        trips() — this is a convenience fallback, not the primary path."""
        s = buid.replace("-", "")
        chunks = [s[i:i + 4] for i in range(0, len(s), 4)]
        return "TRIP" + "-".join(chunks) + "$%010d" % int(trip_id)


if __name__ == "__main__":
    m = MDS()
    d = m.login()
    print("logged in as %s (userId %s, vendorId %s)" % (d["name"], d["userId"], d["vendorId"]))
    buid = sys.argv[1] if len(sys.argv) > 1 else "ivycomptech-IVYHyd"
    g = m.vendor_guids(buid)
    print("%d vendors for %s" % (len(g), buid))
    day = int(datetime.combine(datetime.now().date(), datetime.min.time()).timestamp() * 1000)
    t = m.trips(buid, g, day)
    print("%d trips today, %d unassigned" % (len(t), sum(1 for x in t if not x.get("cabAssigned"))))

File: web/netlify/functions/_mds.mjs
// Shared MDS client for the recommender's functions.
//
// Everything here was verified against production. The two non-obvious things,
// both of which cost real time to find:
//
//   1. `Authorization: <raw token>` — NO "Bearer" prefix. Chrome's HAR export
//      STRIPS Authorization headers, so this header is invisible in .har
//      captures and only appears in "Copy as cURL". Without it every endpoint
//      returns an empty-body 401 from Spring Security, which reads like a bad
//      token rather than a missing header.
//
//   2. /fis/auth/login returns HTTP 200 even on FAILURE, with the real result
//      in `successStatus`. Data endpoints DO use a proper 401. So auth failure
//      has to be detected two different ways depending on the endpoint.
//
//   3. trip/filter rejects more than 10 shifts per call with a 500 — batch.
//
// Env: MDS_EMAIL (or MDS_USERNAME) + MDS_PASSWORD.

export const BASE = process.env.MDS_BASE_URL || 'https://fleet-green.moveinsync.com';

// Module scope survives warm invocations, so a hot function reuses the token
// rather than logging in on every request. Cold starts re-login; that is fine
// (~400 ms) and is why the token is not persisted anywhere.
let cachedToken = null;
let loginInFlight = null;

export function credentials() {
  const username = process.env.MDS_EMAIL || process.env.MDS_USERNAME;
  const password = process.env.MDS_PASSWORD;
  if (!username || !password) {
    const seen = Object.keys(process.env).filter((k) => k.startsWith('MDS_'));
    throw new Error(
      `MDS credentials not configured. Need MDS_EMAIL (or MDS_USERNAME) + MDS_PASSWORD. ` +
      `This function currently sees: ${seen.length ? seen.join(', ') : 'no MDS_* variables at all'}. ` +
      `If you set them in Netlify, check the variable Scope includes "Functions", then redeploy — ` +
      `functions capture environment variables at deploy time, not per request.`);
  }
  return { username, password };
}

async function login() {
  const { username, password } = credentials();
  const res = await fetch(`${BASE}/fis/auth/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
    body: JSON.stringify({ username, password }),
  });
  const json = await res.json();                       // 200 even when rejected
  if (!json.successStatus || !json.data?.token) {
    throw new Error('MDS login rejected: ' + (json.message || 'no token returned') +
                    ' — the MDS password has probably been rotated.');
  }
  const d = json.data;
  return {
    Authorization: d.token,                            // raw, no "Bearer"
    'x-cds-token': d.token,
    user_detail: d.userDetailToken,
    vendor_id: String(d.vendorId),
    Accept: 'application/json, text/plain, */*',
    Referer: `${BASE}/cds-green.html`,
  };
}

// Collapse concurrent logins: wave.mjs fires ~28 parallel calls, and without
// this a cold start would attempt 28 simultaneous logins.
function refreshToken() {
  if (!loginInFlight) {
    loginInFlight = login()
      .then((t) => { cachedToken = t; return t; })
      .finally(() => { loginInFlight = null; });
  }
  return loginInFlight;
}

export async function call(path, body, retried = false) {
  if (!cachedToken) await refreshToken();
  const res = await fetch(BASE + path, {
    method: body ? 'POST' : 'GET',
    headers: body ? { ...cachedToken, 'Content-Type': 'application/json' } : cachedToken,
    body: body ? JSON.stringify(body) : undefined,
  });
  if (res.status === 401 && !retried) {                 // token expired (4h life)
    cachedToken = null;
    await refreshToken();
    return call(path, body, true);
  }
  if (!res.ok) throw new Error(`${path} -> HTTP ${res.status}`);
  return res.json();
}

// Vendor mappings change on the order of months. Caching them removes one
// serial round-trip from every cold call, which matters when the whole handler
// has to finish inside the function timeout.
const guidCache = new Map();
const GUID_TTL = 6 * 3600_000;

export async function vendorGuids(buid) {
  const hit = guidCache.get(buid);
  if (hit && Date.now() - hit.t < GUID_TTL) return hit.guids;
  const r = await call('/fis/vendor/vendorowner/mapping/assigned');
  const m = (r.data?.vendorOwnerMappings || []).find((x) => x.buid === buid);
  if (!m) throw new Error(`no vendor mapping for ${buid} — is this BU assigned to your MDS login?`);
  const guids = m.assignedVendors.map((v) => v.vendorGuid);
  guidCache.set(buid, { t: Date.now(), guids });
  return guids;
}

export async function shifts(buid, guids, dayMs) {
  const r = await call('/fis/cds/shift/v3?nonShift=false&isPlannedTrip=false',
    { dates: [dayMs], buidsAndVendors: [{ businessUnit: buid, vendorGuids: guids }] });
  const out = [];
  for (const d of r || []) {
    for (const b of d.businessUnitOfficeShifts || []) {
      for (const o of b.officeShifts || []) out.push(o);
    }
  }
  return out;
}

// The 'shiftsByOfficeDirectionByBuid' shape is NOT a selectedShifts list — it is
// {buid: [{office:{name,id}, shiftsByDirection:{loginShifts,logoutShifts}}]} and
// `directions` must match whichever list is populated. Getting this wrong is a
// 400; sending more than 10 shifts is a 500.
export async function trips(buid, guids, dayMs) {
  const jobs = [];
  for (const off of await shifts(buid, guids, dayMs)) {
    for (const [direction, key] of [['IN', 'loginShifts'], ['OUT', 'logoutShifts']]) {
      const times = (off[key] || []).map((s) => s.shiftTime);
      for (let i = 0; i < times.length; i += 10) {
        const chunk = times.slice(i, i + 10);
        jobs.push({
          dateList: [dayMs, dayMs],
          businessUnitVendorsList: [{ businessUnit: buid, vendorGuids: guids }],
          directions: [direction],
          shiftsByOfficeDirectionByBuid: {
            [buid]: [{
              office: { name: off.officeName, id: off.officeId },
              shiftsByDirection: {
                loginShifts: direction === 'IN' ? chunk : [],
                logoutShifts: direction === 'IN' ? [] : chunk,
              },
            }],
          },
          selectedShifts: chunk.map((t) => ({
            buid, office: off.officeName, shift: t,
            officeId: off.officeId, direction,
          })),
          plannedTrip: false,
        });
      }
    }
  }
  // Parallel: a large BU is ~28 of these and serial would blow the timeout.
  const results = await Promise.all(jobs.map((b) =>
    call('/fis/cds/trip/filter?source=web&nonShift=false&cabUnassigned=false', b)
      .catch(() => ({ data: [] }))));
  const out = [], seen = new Set();
  for (const r of results) {
    for (const t of r.data || []) {
      if (!seen.has(t.tripId)) { seen.add(t.tripId); out.push(t); }
    }
  }
  return out;
}

// distance=true is REQUIRED: with distance=false, emptyLegInMetres comes back
// as -1 for every cab and the deadhead gate silently stops filtering.
//
// NOTE: no longer the candidate source. Measured on 228 live predictions, this
// list omitted a median 25% of the cabs that regularly run the site (one site:
// 53 regulars, zero offered). Candidates now come from history
// (scorer.candidatesFor); this remains only as a cross-check.
export async function vehicles(buid, tripGuid) {
  const g = encodeURIComponent(tripGuid).replace(/%24/g, '$');
  const r = await call(`/fis/cds/vendor/trips/${buid}/${g}/vehicles?distance=true`);
  return r?.data || [];
}

/**
 * One cab's schedule for a time window — Completed, Ongoing AND future Planned
 * trips, across ALL BUs, each with start/end times and pickup/drop geocodes.
 * Verified live 2026-08-08 (TG-08-V-2063: completed 10:00, ongoing 11:30,
 * planned 14:20 — all in one response).
 *
 * This is the real-time feasibility primitive: everything the old pool's
 * busyVehicle/vehicleNextTripDetails pretended to be, except with locations,
 * dates, and cross-BU visibility.
 *
 * Returns a chain in the exact shape feasible() consumes, sorted by start.
 */
export async function cabChain(registration, startMs, endMs) {
  const r = await call('/fis/cds/trip/cab/trip-history',
    { registration, startTime: startMs, endTime: endMs, allBuids: true });
  const geo = (s) => {
    const [a, b] = String(s || '').split(',').map(Number);
    return Number.isFinite(a) && Number.isFinite(b) ? [a, b] : [null, null];
  };
  const chain = [];
  for (const trips of Object.values(r || {})) {
    for (const t of trips || []) {
      if (!t || !t.startTime || !t.endTime) continue;
      if (t.tripStatus === 'Cancelled') continue;
      const emp = t.employees || [];
      const [slat, slng] = geo(emp[0]?.pickupLoc?.geoCord);
      const [elat, elng] = geo(emp[emp.length - 1]?.dropLoc?.geoCord);
      chain.push({
        start: t.startTime / 1000, end: t.endTime / 1000,
        slat, slng, elat, elng,
        label: `${t.direction || ''} ${t.shift || ''} (${t.buid || '?'})`.trim(),
        tripId: t.tripId, status: t.tripStatus,
      });
    }
  }
  chain.sort((a, b) => a.start - b.start);
  return chain;
}
